accounts/tests.py

```diff
@@ -669,3 +670,79 @@ def test_user_delete_remove_sounds(self):
         self.assertEqual(DeletedSound.objects.filter(user__id=user.id).exists(), True)
 
 
+class UserEmailsUniqueTestCase(TestCase):
+
+    def setUp(self):
+        self.user_a = User.objects.create_user("user_a", password="12345", email='a@b.com')
+        self.user_b = User.objects.create_user("user_b", password="12345", email='c@d.com')
+        self.user_c = User.objects.create_user("user_c", password="12345", email=generate_tmp_email('c@d.com'))
+        SameUser.objects.create(
+            main_user=self.user_b,
+            main_orig_email=self.user_b.email,
+            secondary_user=self.user_c,
+            secondary_orig_email=self.user_b.email,  # Must be same email (original)
+        )
+        # User a never had problems with email
+        # User b and c had the same email, but user_c's was automaitcally changed to avoid duplicates
+
+    def test_redirects_when_shared_emails(self):
+
+        # Try to log-in with user and go to messages page (any login_required page would work)
+        # User a is not in same users table, so redirect should be plain and simple to messages
+        resp = self.client.post(reverse('login'),
+                                {'username': self.user_a, 'password': '12345', 'next': reverse('messages')})
+        self.assertRedirects(resp, reverse('messages'))
+
+        # Now try with user_b. User b had a shared email with user_c, but user_b's email was not changed
+        # Redirect should again be plain and simple
+        resp = self.client.post(reverse('login'),
+                                {'username': self.user_b, 'password': '12345', 'next': reverse('messages')})
+        self.assertRedirects(resp, reverse('messages'))
+
+        # Now user_c. User c had her email changed, therefore will be redirected to the email cleanup page
+        # instead of messages
+        resp = self.client.post(reverse('login'),
+                                {'username': self.user_c, 'password': '12345', 'next': reverse('messages')})
+        self.assertRedirects(resp, reverse('accounts-multi-email-cleanup') + '?next=%s' % reverse('messages'))
+
+        # Now user_c changes his email and tries again, redirect should go to email cleanup page and from there
+        # directly to messages (2 redirect steps)
+        self.user_c.email = 'new@email.com'  # Must be different than generate_tmp_email('c@d.com')
+        self.user_c.save()
+        resp = self.client.post(reverse('login'), follow=True,
+                                data={'username': self.user_c, 'password': '12345', 'next': reverse('messages')})
+        self.assertEquals(resp.redirect_chain[0][0],
+                          reverse('accounts-multi-email-cleanup') + '?next=%s' % reverse('messages'))
+        self.assertEquals(resp.redirect_chain[1][0], reverse('messages'))
```

Comment:
Yeah maybe, but still that sort of "integration" test would be needed in my opinion. The logic for the whole thing is all commented in the `multi_email_cleanup` function because of its potential complexity so I think it is ok there.